Add auto-update on hide workflow for Electron on Mac

nbhargava opened the pull request:
### Details
This change allows the Electron app to auto-update when it has finished downloading the update and it is backgrounded.

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #1011.

### Tests
<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
1. Install and run MinIO locally.
2. Point the `publish` spec in `electron.config.js` at the local server.
3. Publish a new version with the build number set very high (i.e. `1.0.1-999`).
4. Build a new version for local development with a lower build number (i.e. `1.0.1-374`). It is important that this is notarized, so you must use `npm run desktop-build`. This build should also have the `publish` spec pointed at MinIO.
5. Run the app.
6. Verify the app version is `1.0.1-374`.
7. Once the new update has been downloaded, either hide or close the window.
8. The app may reboot a few times.
9. Foreground the app and verify that the app version is `1.0.1-999`.

### Tested On

- [ ] Web
- [ ] Mobile Web
- [X] Desktop
- [ ] iOS
- [ ] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
N/A

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
N/A

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
Video of the experience: https://www.dropbox.com/s/17jhevgcax2xz6x/Screen%20Recording%202021-02-24%20at%2010.13.15%20PM.mov?dl=0

Update modal:
<img width="1312" alt="image" src="https://user-images.githubusercontent.com/993682/109097696-d8b50200-76ed-11eb-9032-9f79e9940ed6.png">



#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
N/A

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
N/A

nbhargava committed:
Add auto-update on hide workflow

Fix double-add

Lint fixes

Add more documentation

Update docs

Move README to desktop/

Show modal prompting user to restart

Use modal to initiate download of update

nbhargava commented:
_**I have read the CLA Document and I hereby sign the CLA**_

roryabraham commented on src/libs/checkForUpdates.js:
```diff
@@ -18,6 +18,9 @@ function checkForUpdates(platformSpecificUpdater) {
     }, UPDATE_INTERVAL);
+
+    // Perform initial check for update
+    platformSpecificUpdater.update();
```
This should be done as part of the `platformSpecificUpdater.init` function. I think this would cause a breaking change on the web platform where the page would refresh in an infinite loop.

roryabraham commented on desktop/main.js:
```diff
@@ -36,6 +36,31 @@ autoUpdater.logger.transports.file.level = 'info';
+            if (!browserWindow.isVisible() && updateDownloaded) {
+                autoUpdater.quitAndInstall();
+            }
+        });
```
In addition to adding event listeners here, you should do an initial check for updates.

roryabraham commented on desktop/main.js:
```diff
@@ -36,6 +36,31 @@ autoUpdater.logger.transports.file.level = 'info';
 // See https://www.npmjs.com/package/electron-log
 Object.assign(console, log.functions);
 
+// Handle auto-update logic
```
This comment isn't very useful - could you expand on what this flag does/means?

roryabraham commented on desktop/main.js:
```diff
@@ -120,7 +149,19 @@ const mainWindow = (() => {
             app.on('before-quit', () => quitting = true);
-            app.on('activate', () => browserWindow.show());
+
+            let firstActivate = true;
```
both this and the next block could benefit from a few comments explaining what they're doing 👍

roryabraham reviewed (changes requested):
Haven't tested this yet, but just had a few comments to start. Overall looks very good!

Also, since this requires and external tool to test, I think it's probably worth including some documentation in the README explaining how to test the desktop auto-updater.

nbhargava commented:
Just pinging this again!

roryabraham commented on README.md:
```diff
@@ -93,6 +93,35 @@ Our React Native Android app now uses the `Hermes` JS engine which requires your
+
+Now, you need to upload a build. Before you can do so, you need to make sure that you can notarize builds. For this you will need an [Apple Developer](https://developer.apple.com) account. Go to the [Certificates, Identifiers, and Profiles](https://developer.apple.com/account/resources/certificates/list) page and create a new certificate for a Developer ID Application. Follow the instructions to  ...
+
+You will need to store your Apple ID in the `APPLE_ID` username and an [app-specific password](https://appleid.apple.com/account/manage) in the `APPLE_ID_PASSWORD` field. Entering your normal password will not work.
```
It's not clear to me where this information should be stored/used. Do you mean in XCode somewhere? In the electron configuration? Could use some more instruction on this step

roryabraham replied:
Oh, I see ... it's used in `desktop/notarize.js`. Looks like those fields can be set as environment variables in the shell session where you run `npm run desktop-build` from, right?

roryabraham replied:
I ended up passing them like this:

```bash
APPLE_ID=<my_apple_id> \
APPLE_ID_PASSWORD=<my_app_specific_password> \
npm run desktop-build
```

roryabraham commented:
I ran into this issue notarizing the desktop app:

```
Package Summary:
 
1 package(s) were not uploaded because they had problems:
	/var/folders/0l/2cyzv7b14fx8rs5w8xpvvbyc0000gn/T/9185A388-AB1B-4C7A-9092-E8BF05A0CA7F/com.expensifyreactnative.chat.itmsp - Error Messages:
		The username rory@seahorsevineyards.com is not a member of the provider . Contact your team admin for assistance. (1296)
2021-02-01 09:16:22.785 altool[10536:83496] *** Error: Unable to notarize app.
2021-02-01 09:16:22.785 altool[10536:83496] *** Error: code 1296 (The username rory@seahorsevineyards.com is not a member of the provider . Contact your team admin for assistance. (1296))
  ```
  
  Did you encounter this issue? Do you know what the problem is/how to resolve the error?

roryabraham reviewed (changes requested):
While I was not completely successful in testing this yet, I did have some suggestions for documentation improvements. I know it's not the best format (sorry), but hopefully you can apply the suggestions from this diff to your PR: [autoUpdaterDocsImprovements.txt](https://github.com/Expensify/Expensify.cash/files/5905592/autoUpdaterDocsImprovements.txt)

nbhargava commented:
I haven't seen that error before so I'm not 100% sure how to resolve it. My best guess is that you have local certificates installed that are for a member of some team/organization and that your personal account, rory@seahorsevineyards.com, is not allowed access to that. If that's the case, my guess would be that switching to your corporate account should let you upload properly.

This issue I found online seems roughly related: https://developer.apple.com/forums/thread/119445

nbhargava commented:
^ real quick on the above: I only tested from a paid developer account, so I'm not sure if that's also a requirement.

roryabraham commented:
Thanks for finding a way to test this and making those documentation updates, but it seems that the only way we can test this is to use a personal paid developer account (we don't want to use our production credentials). So could you please:

1) Move all the great documentation you have here to a separate README in the `desktop` directory (there should be no diff on the README in the root).
2) Make a note in the documentation that in order to test with a notarized build you'll need to have a paid apple developer account.
3) Post a screen recording of the auto-update happening so we can see that it works for you in real time and not just screenshots?

Thanks!

nbhargava commented:
Updated the README as described and below is a video cast of the experience: https://www.dropbox.com/s/un6xgb4gepibmm6/Screen%20Recording%202021-02-02%20at%206.26.57%20PM.mov?dl=0

The flickering on the screen you see right before we re-load it is Big Sur's "verification" of the binary. It is sometimes slow and sometimes this fast.

roryabraham reviewed (dismissed):
Excellent work!

roryabraham commented:
Hi @nbhargava, do you think it would be possible to have the auto-update happen in the middle of the night (even if the laptop is sleeping), rather than when the app is backgrounded?

nbhargava commented:
Shouldn't be too difficult. What counts as middle of the night? 11pm-3am local time?

Also do you want it to be both middle of the night AND backgrounded or just middle of the night?

nbhargava commented:
Oh sorry, missed the part about the laptop sleeping. If the laptop is sleeping I don't know how we'd be able to resume control to initiate the update. Not sure if that changes things.

roryabraham commented:
Hi @nbhargava, after chatting about this internally a bit, we've decided on a change of plans.

Long-term, I think we're going to create a separate issue to investigate remote-hosting the JS bundle so that electron can auto-update its JS without having to close and reopen at all (like the Slack desktop client).

For this PR, could you please update it so that instead of closing and restarting the app when it's hidden, instead we just provide a custom notification to `autoUpdater.checkForUpdatesAndNotify` that includes an `Update Now` button. If the `Update Now` button in the notification is clicked, then it should close the app, install the update, and restart it (no need to immediately hide the app on relaunch).

Thanks!

nbhargava commented:
I'll take a swing at this and see what I can do. The part that may be tricky is handling the fact that the "install" part of `quitAndInstall` doesn't seem to always fire reliably.

nbhargava commented:
@roryabraham -- I've uploaded a new version with the requested changes. Here is a screen recording: https://www.dropbox.com/s/n8a04piq2nijhnv/Screen%20Recording%202021-02-06%20at%201.56.26%20PM.mov?dl=0

roryabraham commented:
Hi @nbhargava, we'd really prefer if the notification and modal maintain a look-and-feel that's consistent with the rest of our application. So that means:

1) Please use a browser notification to inform the user that there's an update available and to provide a button to trigger the update. We have a [LocalNotification](https://github.com/Expensify/Expensify.cash/tree/master/src/libs/Notification/LocalNotification) module which ideally could be extended to have a `showUpdateNotification` function.
2) Please use a `react-native-modal` component to display the confirmation dialog. We have a [Modal](https://github.com/Expensify/Expensify.cash/tree/master/src/components/Modal) module you can use to start.

The modules I linked above use the ES6 module system, so if you want to use them (which would be ideal), you may have to refactor our Electron configuration to enable the importation of ES6 modules into `desktop/main.js`. Do you think that's something you'd be up to? I noticed that [the Electron v12 beta](https://www.electronjs.org/releases/beta?page=4#release-notes-for-v1200-beta1) uses Node 14, and therefore should offer interoperability between CJS and ES6 modules without the use of an external package such as ESM. There is some more info in [this GH issue](https://github.com/electron/electron/issues/21457).

roryabraham commented:
Also, for the record @nbhargava has completed the task we initially gave him, and my previous comment constitutes a change in the scope-of-work. Do we have other examples for what we do when we increase the scope-of-work after the fact?  cc @michaelhaxhiu @laurenreidexpensify @puneetlath – have you guys encountered this situation yet?

puneetlath commented:
I think in that case let's just increase the scope of the Upwork job/pay by the same scale as the increase in scope.

michaelhaxhiu commented:
+1 to that @puneetlath. Let's provide a bonus for the additional scope if it's agreeable for @nbhargava !

nbhargava commented:
Sounds great -- will take a swing at this and report back when I have an update!

nbhargava commented:
Also to clarify before I dive too deep, is the "upgrade to electron v12 beta" the preferred path you'd have me take on this, or is something else preferable?

roryabraham commented:
Yeah, I think the ideal implementation of this would use our existing modules for LocalNotifications and Modals, but I don't want to see those converted to use CJS. So that means that ideally you can configure `desktop/main.js` to import and use ES6 modules. I think there are two main ways you could do that:

1) Using the [esm](https://www.npmjs.com/package/esm) package. (I tried this a while back without success, but it _should_ be possible).
2) Using Node 13+ (aka upgrade Electron to the beta).

I think I would prefer to just upgrade Electron, but I'll let you dig into it and report back with which option you think will be best. The best source of information I've found on this topic so far is [this GH issue](https://github.com/electron/electron/issues/21457).

tgolen commented:
Hi! I just wanted to chime in with a possible alternative solution that might help solve the issue of importing ES6 modules into `desktop/main.js`. 

Rather than importing code into that file to try and display a modal (and having to also import all of React), I think I would prefer if we just keep all the UI elements as part of the action app (ie. have it all be in the same bundle as `/src`). Then in `desktop/main.js`, whenever we need to display the upgrade modal, there should be a way (sorry, I am not familiar with Electron specifically) to just signal the app to display the upgrade modal? Like through some kind of event system, or maybe even with Onyx.

roryabraham commented:
Thanks @tgolen, great point about trying to keep React-specific UI code (i.e: the `Modal` module) out of Electron's main process.

I _think_ you should be able to use the `LocalNotification` module in there without using React, though, because it just uses browser notifications, which should be natively supported by Electron.

Check out [this example](https://github.com/Expensify/Expensify.cash/blob/ef2cf997bc91a327186ddb4c23d3a5ad7bff327c/src/libs/Notification/LocalNotification/BrowserNotifications.js#L106) of a function exported from the `LocalNotification` library. You'll notice that you can provide an `onClick` callback for a browser notification. Ideally you can use that callback to signal the app to display the upgrade modal as @tgolen has suggested.

Does that make sense? Any questions or concerns?

nbhargava commented:
Definitely on-board for keeping react code out of `desktop/main.js`. This path forward makes sense. I'll keep poking around on this but may not have an updated PR until this weekend.

nbhargava commented:
After playing around for a while, it seems like Electron 12 still doesn't seem to have ES6 module support fully turned on. I was able to make some progress with 'esm', but we'll see if I run into any more pitfalls.

As part of trying to drive through imports of LocalNotification, I'm running into issues in upstream dependencies. I filed this PR if someone could please take a look! https://github.com/Expensify/expensify-common/pull/334

tgolen commented:
Why are we still trying to import ES6 modules into that file? The above solution I mentioned was intended to remove the need to do this at all. Specifically, what modules are you importing?

nbhargava commented:
At the time, I was trying to import `LocalNotification`, but as you mention, it's much cleaner to just use Electron's IPC to get this done.

FWIW, in the future, the `esm` strategy does work quite well if you do need to end up importing ES6 modules in the future.

nbhargava commented:
This new diff has the updates that pop open a modal when the download is available and then triggers the download when the user interacts with the popover. Below is a video of the interaction:

https://www.dropbox.com/s/6oa7p4qtnms58yr/Screen%20Recording%202021-02-21%20at%204.19.05%20PM.mov?dl=0

nbhargava committed:
Remove electron-is-dev

nbhargava committed:
Fix package-lock

roryabraham commented:
@nbhargava The video you've provided looks great to me! @AndrewGable, what do you think?

However, I noticed your code has several lint and test failures, so could you please work on resolving those before I dive into a review?

nbhargava commented:
Yup, I realized yesterday that there are some rebase issues due to refactors that happened while I was developing. 

Will update later today and re-ping the thread once those are fixed.

AndrewGable commented:
Looks great to me, cc @Expensify/design to make sure this is the UI element we'd like.

shawnborton commented:
Are the screenshots on the original comment correct? From my understanding of reading through this, it seems like there should be some kind of Update button but I wasn't seeing one in the screenshots. It also looks like the video link you provided was deleted.

AndrewGable commented:
I believe [this is the latest video](https://www.dropbox.com/s/6oa7p4qtnms58yr/Screen%20Recording%202021-02-21%20at%204.19.05%20PM.mov?dl=0)

shawnborton commented:
Got it, @nbhargava can you update the original comment with both the video as well as static screenshots that clearly show the update flow? 

In terms of the UI, I think it's great and is the best thing we can do right now before we have some kind of in-app notification/growl/toast component created. We can address that as a separate problem/solution when we start to develop the need for it.

nbhargava committed:
Fix rebase issues

nbhargava commented:
Updated the original comment with the video & a static screenshot of the modal and fixed all the rebase issues

Not sure how to re-run the iOS tests, but from the log output: `This is an error with npm itself`, so I'm not too sure that I'm newly causing this error?

roryabraham commented on desktop/main.js:
```diff
@@ -137,7 +181,12 @@ const mainWindow = (() => {
-            app.on('activate', () => browserWindow.show());
+
+            // Hide the app if we expected to upgrade to a new version but never did.
+            if (expectedUpdateVersion !== undefined && app.getVersion() !== expectedUpdateVersion) {
```
NAB suggestion

```suggestion
            if (expectedUpdateVersion && app.getVersion() !== expectedUpdateVersion) {
```

roryabraham replied:
just in case you weren't familiar: `NAB` === `Not a blocker`

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,94 @@
+import styles from '../../styles/styles';
+import themeColors from '../../styles/themes/default';
+
+/**
```
Did you mean to leave this comment here?

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,94 @@
+                    onSubmit={this.submitAndClose}
+                    onClose={() => this.setState({isModalOpen: false})}
+                    isVisible={this.state.isModalOpen}
+                    backgroundColor={themeColors.componentBG}
```
I think this prop is missing from the `ModalPropTypes` props validation.

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,94 @@
+                    {this.props.onConfirm && (
+                        <TouchableOpacity
+                            style={[styles.button, styles.buttonSuccess, styles.buttonConfirm]}
+                            underlayColor={themeColors.componentBG}
```
underlayColor is not a valid prop of `TouchableOpacity`. It is a prop of `TouchableHighlight`, but we don't use that component anywhere.

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,94 @@
+
+const propTypes = {
+    // Optional callback to fire when we want to trigger the update.
+    onConfirm: PropTypes.func,
```
I think we should rename this from `onConfirm` to `onSubmit`, to match the underlying modal prop.

roryabraham commented on src/libs/Notification/LocalNotification/BrowserNotifications.js:
```diff
@@ -117,4 +117,15 @@ export default {
+    pushUpdateAvailableNotification() {
+        push({
+            title: 'Update available',
+            body: 'A new version of Expensify.cash is available for download.',
```
Is it "available for download", or is it already downloaded?

```suggestion
            body: 'A new version of Expensify.cash is available!',
```

roryabraham commented on src/libs/Notification/LocalNotification/BrowserNotifications.js:
```diff
@@ -117,4 +117,15 @@ export default {
+    /**
+     * Create a notification to indicate that an update is available.
+     */
+    pushUpdateAvailableNotification() {
```
Let's update this so that it takes a `version` parameter, then add an `onClick` handler that does:

```javascript
Onyx.merge(ONYXKEYS.UPDATE_VERSION, version);
```

roryabraham commented on src/setup/index.desktop.js:
```diff
@@ -1,8 +1,18 @@
     });
+
+    ipcRenderer.on('update-downloaded', (_, version) => {
+        LocalNotification.showUpdateAvailableNotification();
```
```suggestion
        LocalNotification.showUpdateAvailableNotification(version);
```
Then get rid of the next line that immediately updates Onyx. This way, the modal only appears when the user clicks the notification.

nbhargava replied:
I had started with this approach, but the problem I realized when testing locally is that if a user has turned off (or possibly never enabled) notifications for the app, then they will never see this.

roryabraham replied:
> I had started with this approach, but the problem I realized when testing locally is that if a user has turned off (or possibly never enabled) notifications for the app, then they will never see this.

Oh, interesting, I hadn't considered that. In that case, I think we would want to just add an `Update` button in the settings page as a backup if there is an update available. You could add that yourself, or we can also just implement that in a follow-up PR, but either way I'd prefer to see the modal only be displayed when the user clicks the notification.

nbhargava replied:
I'm going to vote making that a follow-up PR!

roryabraham reviewed (changes requested):
Overall code looks good, just a few minor changes. I re-ran the iOS tests and they passed, so you're good there.

Also, I noticed that the `UpdateAppModal` module has an `index.native.js` and `index.website.js` that are identical. We just fixed a webpack bug yesterday so that you shouldn't need to do that anymore. Just put your desktop-specific code in `index.desktop.js` and the default in `index.js` and it should work.

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,94 @@
+        return (
+            <>
+                <Modal
+                    type={CONST.MODAL.MODAL_TYPE.RIGHT_DOCKED}
```
Up to you, but you should see what happens when you leave this `type` prop out entirely. That will cause the modal to be centered in the screen rather than right-docked, which to me feels maybe more appropriate for this modal.

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,94 @@
+                            >
+                                Update App
+                            </Text>
+                        </TouchableOpacity>
```
NAB, but might be nice to have a "Maybe Later" cancel button too.

nbhargava committed:
Respond to reviewer feedback

nbhargava commented:
Updated with your feedback and added a new video/screenshot to the top.

roryabraham commented:
@nbhargava sorry, it seems like the video link is broken.

roryabraham commented on src/components/UpdateAppModal/BaseUpdateAppModal.js:
```diff
@@ -0,0 +1,83 @@
+
+    // Version string for the app to update to.
+    // eslint-disable-next-line react/no-unused-prop-types
+    version: PropTypes.string,
```
Okay, actually I see why you did this. NAB but what I would've done is:

1) Create a new file called `src/components/UpdateAppModal/UpdateAppModalPropTypes`
2) Export the props types and default props you have in here as named exports like `export {propTypes, defaultProps}` from that file
3) In `UpdateAppModal/index.js` and `UpdateAppModal/desktop.index.js`, use the prop types from step 2, unaltered. Then just pass the one prop that's needed in `BaseUpdateAppModal` – `onSumbit`
4) In `BaseUpdateAppModal`, set the `propTypes` to `_.omit(propTypes, 'version')`. Same for default props.

roryabraham commented:
@nbhargava Any update here?

nbhargava committed:
Update props

nbhargava commented:
Sorry, forgot to push! The video is working for me, even in an incognito window. Are you still having issues with it?

roryabraham reviewed (approved):
LGTM and great work 👍

marcaaron commented:
`propTypes` warnings were not fixed :( we should just make them errors

roryabraham commented:
> propTypes warnings were not fixed

Idk why I didn't think about this some more (should have held on merging), but it seems like a false warning to me. The prop types were imported from another file then altered, but should actually include `onSubmit`.
